### DISCUSSION TITLE
Status buffer jumping

### DISCUSSION THREAD

**Discussion by xendk** [magit/magit#4515] [Support] [answered]
I have a pet peeve...

Whenever I stage a hunk of a file, the point moves around in ways I haven't quite figured out. But the thing that really gets my goat me is that the buffer scrolls to the top, which is very annoying, as I've often scrolled down to show as much of the diffs in the "Unstaged changes" area.

I would prefer the view to stay where it is and the point just to move up to the beginning of the next hunk, at least for small hunks (staging in the middle of a big hunk that covers the whole screen is bound to reposition the view).

Is it just me that's messed my setup up, or does the current behavior seem natural to everybody else?

**Comment by xendk** [comment]
And after posting, I do notice that when having most hunks shown and being so far down the list that I only see diffs, it repositions so the filename of the file I just staged a hunk in is in the center.

But most of the time I have just a few hunks, and starting from the top, so I experience the "top-of-buffer" issue. And with a good chunk of untracked files, that often means that most of "Unstaged changes" ends up below the last visible line.

**Comment by tarsius** [comment, marked as answer]
Point is supposed to be put on another close by hunk or file.

If I remember there was a bug that defeated that, which was only fixed somewhat recently. This year, or maybe last. What version are you using?

**Reply by xendk** [reply]
OK, had to update first now I'm on `Magit v3.3.0-17-g348d9b98, Git 2.30.2, Emacs 27.1, gnu/linux`...

Initially it seems... Better... But.. Positioning the buffer so that the first hunk is at the top of the window and staging the second one, point does jump to the following hunk, but it still scrolls the buffer so that the first line of the buffer is visible, which puts the point about one tenth above the bottom of the window (lots of untracked files in this repo). I would kinda expect the point to keep its vertical position in the window even though the buffer content changes.

**Reply by tarsius** [reply]
IMO the limited jumping that still occurs in some cases is okay.

**Reply by xendk** [reply]
Limited? I wouldn't call it limited, the point goes from the top of the window to the bottom and 90% of the window content gets replaced. :grin: But does magit move the viewport, or  is it just moving the point and something else is taking the decision that the `Head:` line is the most important thing to show (it isn't)?

**Reply by tarsius** [reply]
> Limited? I wouldn't call it limited, the point goes from the top of the window to the bottom and 90% of the window content gets replaced.

Yeah, that doesn't happen for me. Or only in very limited circumstances.

> But does magit move the viewport, or is it just moving the point and something else is taking the decision that the Head: line is the most important thing to show (it isn't)?

Magit completely erases the content of the buffer, then it recreates it from scratch. That isn't optimal for obvious performance reasons and is why it don't want to invest too much time optimizing the following instead of changing the former. Then it locates the section that represents the same thing as the previously selected section, or at least a close by section and moves point and focuses on that.

**Reply by xendk** [reply]
> Yeah, that doesn't happen for me. Or only in very limited circumstances.

You're a tidy person that doesn't have a lot of untracked files lying about (I'm obviously not)? But I just remembered `i p`, that might improve things.

> it don't want to invest too much time optimizing

Makes sense.

> and moves point and focuses on that

How does it focus? Maybe I can monkeypatch some temporary solution if I know what to grep for.

**Reply by tarsius** [reply]
Look at functions on `magit-section-goto-successor-hook`.

And `magit-section-movement-hook`; heck maybe all you need is `(remove-hook 'magit-section-movement-hook #'magit-hunk-set-window-start)`), or not. Worth a try.

**Reply by xendk** [reply]
The remove-hook didn't quite do it, but there's something to work with there. Thanks!
